gemini/use-cases/retrieval-augmented-generation/developer_code_chat/utils/intent_routing.py

```diff
@@ -0,0 +1,401 @@
+# Copyright 2023 Google LLC. This software is provided as-is, without warranty
+# or representation for any use or purpose. Your use of it is subject to your
+# agreement with Google.
+"""Generative AI Models based functions to respond questions"""
+# Imports
+import configparser
+import json
+import logging
+
+from vertexai.generative_models import GenerationConfig
+
+
+class IntentRouting:
+    """genai Assistant"""
+
+    def __init__(
+        self, config_file: str = "config.ini", logger=logging.getLogger()
+    ) -> None:
+        self.config = configparser.ConfigParser()
+        self.config.read(config_file)
+        self.logger = logger
+
+        self.genai_qna_parameters = GenerationConfig(
+            temperature=float(self.config["genai_qna"]["temperature"]),
+            max_output_tokens=int(self.config["genai_qna"]["max_output_tokens"]),
+            top_p=float(self.config["genai_qna"]["top_p"]),
+            top_k=int(self.config["genai_qna"]["top_k"]),
+        )
+        self.genai_chat_parameters = GenerationConfig(
+            temperature=float(self.config["genai_chat"]["temperature"]),
+            max_output_tokens=int(self.config["genai_chat"]["max_output_tokens"]),
+        )
+
+    def get_chat_history(self, chat):
+        """Get Chat History"""
+
+        chat_history = chat.history
+
+        if len(chat_history) == 0:
+            return []
+        return chat_history[-6:]
+
+    def other_intent(self, text, chat_model):
+        """Respond for Other type of Intent"""
+
+        chat = chat_model.start_chat()
+        response = chat.send_message(
+            f"""{text}""", generation_config=self.genai_chat_parameters
+        )
+        message = response.text
+        return message
+
+    def greetings(self, chat_model, text):
+        """Respond for Greetings or Welcome Intent"""
+
+        enabled_programming_language = self.config["default"][
+            "enabled_programming_language"
+        ].split(",")
+        enabled_programming_language = ", ".join(
+            [lang.title() for lang in enabled_programming_language]
+        )
+        enabled_qna_programming_language = self.config["default"][
+            "enabled_qna_programming_language"
+        ].split(",")
+        enabled_qna_programming_language = ", ".join(
+            [lang.title() for lang in enabled_qna_programming_language]
+        )
+
+        chat = chat_model.start_chat()
+        response = chat.send_message(
+            f"""You are Generative AI powered genai Learning Assistant.
+
+        You are trained ONLY to answers questions related to following programming languages: {enabled_programming_language}
+
+        Write a brief greeting message:"""
+        )
+        # parameters_local = copy.deepcopy(self.genai_chat_parameters)
+        # parameters_local["temperature"] = 1
+        parameters_local = GenerationConfig(
```

Comment:
Why do you need this extra generation config since you already have one in the class?